lnd.go

```diff
@@ -644,65 +643,69 @@ func waitForWalletPassword(grpcEndpoint, restEndpoint string,
 
-	// Start a gRPC server listening for HTTP/2 connections, solely
-	// used for getting the encryption password from the client.
-	lis, err := net.Listen("tcp", grpcEndpoint)
-	if err != nil {
-		fmt.Printf("failed to listen: %v", err)
-		return nil, nil, err
-	}
-	defer lis.Close()
+	// Use a WaitGroup so we can be sure the instructions on how to input the
+	// password is the last thing to be printed to the console.
+	var wg sync.WaitGroup
 
-	// Use a two channels to synchronize on, so we can be sure the
-	// instructions on how to input the password is the last
-	// thing to be printed to the console.
-	grpcServing := make(chan struct{})
-	restServing := make(chan struct{})
+	for _, grpcEndpoint := range grpcEndpoints {
+		// Start a gRPC server listening for HTTP/2 connections, solely
+		// used for getting the encryption password from the client.
+		lis, err := net.Listen("tcp", grpcEndpoint)
+		if err != nil {
+			ltndLog.Errorf("password RPC server unable to listen on %s",
+				grpcEndpoint)
+			return nil, nil, err
+		}
+		defer lis.Close()
 
-	go func(c chan struct{}) {
-		rpcsLog.Infof("password RPC server listening on %s",
-			lis.Addr())
-		close(c)
-		grpcServer.Serve(lis)
-	}(grpcServing)
+		wg.Add(1)
+		go func() {
+			rpcsLog.Infof("password RPC server listening on %s", lis.Addr())
+			wg.Done()
+			grpcServer.Serve(lis)
+		}()
+	}
 
 	// Start a REST proxy for our gRPC server above.
 	ctx := context.Background()
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
 	mux := proxy.NewServeMux()
-	err = lnrpc.RegisterWalletUnlockerHandlerFromEndpoint(ctx, mux,
-		grpcEndpoint, proxyOpts)
+
+	err := lnrpc.RegisterWalletUnlockerHandlerFromEndpoint(ctx, mux,
+		grpcEndpoints[0], proxyOpts)
 	if err != nil {
 		return nil, nil, err
 	}
+
 	srv := &http.Server{Handler: mux}
 	defer func() {
 		// We must shut down this server, since we'll let
 		// the regular rpcServer listen on the same address.
 		if err := srv.Shutdown(ctx); err != nil {
-			ltndLog.Errorf("unable to shutdown gPRC proxy: %v", err)
+			rpcsLog.Errorf("unable to shutdown password gRPC proxy: %v", err)
 		}
 	}()
 
-	go func(c chan struct{}) {
-		listener, err := tls.Listen("tcp", restEndpoint,
-			tlsConf)
+	for _, restEndpoint := range restEndpoints {
+		lis, err := tls.Listen("tcp", restEndpoint, tlsConf)
 		if err != nil {
-			ltndLog.Errorf("gRPC proxy unable to listen "+
-				"on localhost%s", restEndpoint)
-			return
+			ltndLog.Errorf("password gRPC proxy unable to listen on %s",
+				restEndpoint)
+			return nil, nil, err
 		}
-		rpcsLog.Infof("password gRPC proxy started at "+
-			"localhost%s", restEndpoint)
-		close(c)
-		srv.Serve(listener)
-	}(restServing)
-
-	// Wait for gRPC and REST server to be up running.
-	<-grpcServing
-	<-restServing
+
+		wg.Add(1)
+		go func() {
+			rpcsLog.Infof("password gRPC proxy started at %s", lis.Addr())
+			wg.Done()
+			srv.Serve(lis)
+		}()
+	}
+
+	// Wait for gRPC and REST servers to be up running.
+	wg.Wait()
```

Comment:
👍  for swapping out the old channel based synchronization with the usage of wait groups here.